Applications/spire/source/ui/security_widget.cpp

```diff
@@ -1,5 +1,123 @@
 #include "spire/ui/security_widget.hpp"
+#include <QKeyEvent>
+#include "spire/security_input/security_input_dialog.hpp"
+#include "spire/spire/dimensions.hpp"
 
 using namespace Beam;
+using namespace boost::signals2;
 using namespace Nexus;
 using namespace Spire;
+
+SecurityWidget::SecurityWidget(Ref<SecurityInputModel> input_model,
+    Theme theme, QWidget* parent)
+    : QWidget(parent),
+      m_input_model(input_model.Get()) {
+  setFocusPolicy(Qt::StrongFocus);
+  if(theme == Theme::DARK) {
+    setStyleSheet("background-color: #D0D0D0;");
+  } else {
+    setStyleSheet("background-color: #FFFFFF;");
+  }
+  m_layout = new QVBoxLayout(this);
+  m_layout->setContentsMargins({});
+  m_empty_window_label = std::make_unique<QLabel>(
+    tr("Enter a ticker symbol."), this);
+  m_empty_window_label->setAlignment(Qt::AlignCenter);
+  m_empty_window_label->setStyleSheet(QString(R"(
+    font-family: Roboto;
+    font-size: %1px;
+    padding-top: %2px;)").arg(scale_height(12)).arg(scale_height(16)));
+  m_layout->addWidget(m_empty_window_label.get());
+}
+
+void SecurityWidget::set_widget(QWidget* widget) {
+  m_empty_window_label.reset();
+  m_widget = widget;
+  m_layout->addWidget(m_widget);
+  m_widget->show();
+  m_widget->installEventFilter(this);
+}
+
+connection SecurityWidget::connect_change_security_signal(
+    const ChangeSecuritySignal::slot_type& slot) const {
+  return m_change_security_signal.connect(slot);
+}
+
+bool SecurityWidget::eventFilter(QObject* object, QEvent* event) {
+  if(object == m_widget && event->type() == QEvent::KeyPress) {
+    keyPressEvent(static_cast<QKeyEvent*>(event));
+    return event->isAccepted();
+  }
+  return QWidget::eventFilter(object, event);
+}
+
+void SecurityWidget::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_PageUp) {
+    if(m_current_security != Security()) {
+      auto s = m_securities.push_front(m_current_security);
+      if(s != Security()) {
+        m_current_security = s;
+        m_change_security_signal(s);
+      }
+    }
+    event->accept();
+  } else if(event->key() == Qt::Key_PageDown) {
+    if(m_current_security != Security()) {
+      auto s = m_securities.push_back(m_current_security);
+      if(s != Security()) {
+        m_current_security = s;
+        m_change_security_signal(s);
+      }
+    }
+    event->accept();
+  }
+  auto pressed_key = event->text();
+  if(pressed_key[0].isLetterOrNumber()) {
+    auto dialog = new SecurityInputDialog(Ref(*m_input_model), pressed_key,
```

Comment:
This introduces a memory leak.

Reply:
I have the [WA_DeleteOnClose](http://doc.qt.io/qt-5/qt.html#WidgetAttribute-enum) set on the dialog, is there something else I'm missing?